fix(ember-data): improve belongsTo and hasMany relationship typing

makepanic opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:

  * `remoteType` return type: [hasMany](https://github.com/emberjs/data/blob/v3.14.0/packages/store/addon/-private/system/references/has-many.js#L76-L81), [belongsTo](https://github.com/emberjs/data/blob/v3.14.0/packages/store/addon/-private/system/references/reference.ts#L76-L82) (Note the ember-data type also declares `identity` but it doesn't seem to be used)
  * I made BelongsToReference and HasManyReference generic (HasManyReference had some generic parts already). We might be able to narrow the generic type further by asserting that `T extends Model` but I'm not sure if it's too strict.
  * `hasMany` and `belongsTo` now extracts the type by extracting it from the instance. This can create false-positives as it can't decide if a field is an attribute or a relationship. This drawback [already existed previously](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ember-data/index.d.ts#L26).
  * __Note__: this also adjusts the tsconfig to enable checking of `belongsTo` and `hasMany` when used as a decorator

- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

makepanic commented:
as mentioned in discord, I'd also be ok with removing the "cleverness" and omit automatic type extraction. 

Having a generic `BelongsToReference` would make things easier than they currently are

mike-north commented on types/ember-data/index.d.ts:
```diff
@@ -546,11 +558,11 @@ export namespace DS {
          * Get the reference for the specified belongsTo relationship.
          */
-        belongsTo(name: RelationshipsFor<this>): BelongsToReference;
+        belongsTo<T extends keyof this = RelationshipsFor<this>>(name: T): BelongsToReference<UnpackComputedProperty<this[T]>>;
```
Unfortunately `this` used as a typeparam has been a source of _significant breaking changes_  to our types as new compiler versions are released. The fact that it includes `protected` and `private` members, when used internally, but those _should not be visible externally_ results in some significant nuance that few other type packages seem to encounter (i.e., the ember types have been one of only 5-6 packages that break routinely around compiler changes that affect `this`)

As the person who often has to urgently dig us out of these kinds of problems, I'd prefer to 

(a) not use `<this>` or
(b) be ready to, at a moments notice, eject out to a _less clever_ typing in the event a new TS nightly build starts to block our CI pipeline

makepanic replied:
What do you think about removing the `this` usage for belongsTo/hasMany and simply make HasManyReference and BelongsToReference generic?

I also hit some issues with the old `RelationshipsFor<this>` 

e.g.
```
belongsTo<T extends DS.Model>(name: string): BelongsToReference<T>;
hasMany<T extends DS.Model>(name: string): HasManyReference<T>;
```

mike-north replied:
I'd consider this a possible misuse of generics. Generally they're intended to be used as a tool for expressing a relationship between two or more types (expressed in terms of one or more typeparams), and in the case above you use it in only once.

What you have here is just a casting mechanism where we'd have to provide a typeparam at every usage site. If we're going to encourage this kind of casting, I'd prefer that it _looks like_ casting (i.e., using the `as` keyword)

From a type safety standpoint,
```ts
belongsTo<T extends DS.Model>(name: string): BelongsToReference<T>;

this.belongsTo<Book>('book');
```
is no more type safe than 
```ts
belongsTo(name: string): BelongsToReference<DS.Model>;

this.belongsTo('book') as Book;
```

makepanic replied:
Yes that makes sense. 

For us there were multiple places where the current api is really verbose if one wants to type it. e.g. the following is needed right now:

```
const ref = model.belongsTo('rel') 
ref.load() as Promise<Rel | null>;
ref.value() as Rel | null;
ref.reload() as Promise<Rel | null>;
```

Which could be changed to:
```
const ref = model.belongsTo('rel') as BelongsToRef<Rel>;
ref.load() // Promise<Rel | null>;
ref.value() // Rel | null;
ref.reload() // Promise<Rel | null>;
```

Which would be possible if we make `BelongsToReference` generic in the same way `HasManyReference` is currently generic. 

Would you be ok with that?

mike-north commented on types/ember-data/test/belongs-to.ts:
```diff
@@ -7,11 +7,17 @@ class Folder extends DS.Model {
 declare module 'ember-data/types/registries/model' {
     export default interface ModelRegistry {
         folder: Folder;
+        owner: Owner;
```
Could we name this something else? I had to hunt around to make sure that this wasn't the `owner` from RFC-232 tests (i.e., `this.owner` (ApplicationInstance) )

makepanic replied:
Sure. I've renamed it.

makepanic commented:
I've updated the PR and removed all the type inference.

Now the PR only changes BelongsToReference to be generic in the same way HasManyReference already was.

sandersn commented:
@mike-north mind taking another look? I kind of agree with you that a type parameter on BelongsToReference is suspicious, but @makepanic has a good point that the current code leads to a lot of casting. But the decision to merge or not is up to you.

chriskrycho commented:
cc. also @jamescdavis
